### PR TITLE
mpdclient module for controlling MPD servers

### PR DESCRIPTION
**Pull request by glueckself**
Hi!

This is my first module, I hope it's ok to use this draft pull request for feedback.

This code is untested for two reasons: I'm not at home (for the next three weeks), but more importantly (and this brings me to the questions):

* How do I install a module that is not in the official repository? Is it enough to just copy it into the module directory in ProjectAlice and do a sync+download via alice cli?
* How do I enable/disable intents? I think it would be good to enable/disable them based on the state of the player (no point in stopping twice).
* How do I integrate with other music modules (e.g. YoutubeJukebox)? It also uses rather "simple" dialogs ("play music of"). Do we need a, let's call it "MusicControl", module that has the dialogs and talks and uses backends (mpd, youtube, spotify, ...)?
* Do I put the configuration in config.py? How does configuration work when installing via alice cli?

Mistakes I know I made:
* I started by copying another module. Should I redo my module with Moduler or does this seem good enough to go on?
* Small number of utterances. I'll write more.

Please let me know what else I missed :) Thanks!

### CODE REVIEW

**Comment by Psychokiller1888** [comment]
First of all, thank you for your contribution!

> How do I install a module that is not in the official repository? Is it enough to just copy it into the module directory in ProjectAlice and do a sync+download via alice cli?

For now, manually only... It is planned but not yet implemented. You have to upload your module inside ProjectAlice's modules directory and edit "config.py" at the root of ProjectAlice and add your module information in there, then restart Alice

**Comment by Psychokiller1888** [comment]
> How do I enable/disable intents? I think it would be good to enable/disable them based on the state of the player (no point in stopping twice).

MqttManager can enable/disable intents on the fly with the "configureModuleIntents()" which is available in alpha 2, due to release today or tomorrow

**Comment by Psychokiller1888** [comment]
> How do I integrate with other music modules (e.g. YoutubeJukebox)? It also uses rather "simple" dialogs ("play music of"). Do we need a, let's call it "MusicControl", module that has the dialogs and talks and uses backends (mpd, youtube, spotify, ...)?

This is what Alice does for every telemetry data, such as temperature, humidity, light etc etc. There's one module with the intents that store all these informations. To answer the question, it is true that a Youtube music app will conflict with yours and vice versa. We have two things on the burner, the support for "Play some Nirvana **with the {youtube} app** and "Open {youtube app}" "Play some nirvana"

Past that, I haven't gone deeper into thinking it. A central would be cool, but where to stop? I mean, it's like news, you have many different news site, and we'd need a central module for it too... And many other things. This needs proper thinking

**Comment by Psychokiller1888** [comment]
> Do I put the configuration in config.py? How does configuration work when installing via alice cli?

The configurations must be distributed through your module's config.dist.json. This file is copied and then used by Alice for the module's settings, avoid anybody to push his secrets in an uncleaned config file. The file is kept when updating modules. Configuring through cli is not yet made and requires manual editing of the file for now. Configuring through the interface (alpha 2) is possible and works

File: PublishedModules/glueckself/mpdclient/dialogTemplate/de.json
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/dialogTemplate/de.json]: Your intents are all disabled by default? This means they will never trigger, is that intended?

File: PublishedModules/glueckself/mpdclient/mpdclient.install
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.install]: This means people that allow Alice to use internet services won't be able to use your module, I guess that's not intended?

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: single quotes please :)

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: fstrings need an 'f' at the start =>

`self._logger.warn(f'[{self.name}] MPD host not configured, not doing anything.')`

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: self._mdpConnected = True

File: PublishedModules/glueckself/mpdclient/mpdhelper.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdhelper.py]: If you are distributing the API, you don't need to specify it in the module install file. Having it specified in the install file will make Alice download the api and install it

File: PublishedModules/glueckself/mpdclient/talks/en.json
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/talks/en.json]: typo

File: PublishedModules/glueckself/mpdclient/dialogTemplate/de.json
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/dialogTemplate/de.json]: Yes, this is intended because it they should be enabled depending on the mpd state.
When the client is disconnected, none of the intents make sense (maybe only a try-to-connect-intent, but I prefer that the module keeps trying to connect itself).

Comment (Psychokiller1888) [reply PublishedModules/glueckself/mpdclient/dialogTemplate/de.json]: Ok, I see, good thing! When connected, you can use MqttManager.configureIntents() (alpha2)

File: PublishedModules/glueckself/mpdclient/mpdhelper.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdhelper.py]: I copied the from Sonata, a GUI MPD client. Am I right that I need to keep the copyright header?

File: PublishedModules/glueckself/mpdclient/mpdclient.install
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.install]: No it's not, thank you.
I thought that meant that the module requires online connection (stayCompletlyOffline has to be false). I'll fix that later today.

File: PublishedModules/glueckself/mpdclient/mpdhelper.py
Comment (Psychokiller1888) [reply PublishedModules/glueckself/mpdclient/mpdhelper.py]: You HAVE TO keep the copyrights :) I just checked, it's under GNU, so you can redistribute it yes. So remove the requirements in the install and readme file so Alice won't install it again :)

File: PublishedModules/glueckself/mpdclient/dialogTemplate/de.json
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/dialogTemplate/de.json]: Thanks! :)

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: Sorry, habit, fixed.
Should I make sure that mpdhelper.py also conforms to the code guidelines?
As noted below, that is a file that I copied from another project.

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: fixed, thanks

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: fixed, sorry.

File: PublishedModules/glueckself/mpdclient/talks/en.json
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/talks/en.json]: fixed, thank you.

**Review by Psychokiller1888 — COMMENTED** [review]
Don't rewrite the api, it follows python guidelines and it would become a pain to update in case needed.

File: PublishedModules/glueckself/mpdclient/mpdclient.install
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.install]: This means the user HAS to allow internet usage, MPD is local no? In which case you don't need that condition at all, we don't care if the user is online, offline, allows or forbids connectivity etc etc

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: I think typo here? "pass**wod**"

File: PublishedModules/glueckself/mpdclient/mpdhelper.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdhelper.py]: The requirements are for the base mpd library.
It goes python-mpd2 (via requirements) - mpdhelper.py - mpdclient.py

I don't really understand how to use the mpd2 library, even after reading mpdhelper.py. To be honest, I also don't understand how the mpdhelper.MPDClient class provides the play()/stop()/... methods, but I've seen them used in Sonatas code (I guess they are added with setattr?). That's why I'm using the helper.

Comment (Psychokiller1888) [reply PublishedModules/glueckself/mpdclient/mpdhelper.py]: Ooh, sorry, did not understand it that way! So it's ok :)

File: PublishedModules/glueckself/mpdclient/mpdclient.install
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.install]: I get it now. Thank you!
MPD can local or remote. I guess it will be usually in the same network as Alice (my case) or maybe a hop or two away.

**Comment by glueckself** [comment]
> This is what Alice does for every telemetry data, such as temperature, humidity, light etc etc. There's one module with the intents that store all these informations. To answer the question, it is true that a Youtube music app will conflict with yours and vice versa. We have two things on the burner, the support for "Play some Nirvana **with the {youtube} app** and "Open {youtube app}" "Play some nirvana"
>
> Past that, I haven't gone deeper into thinking it. A central would be cool, but where to stop? I mean, it's like news, you have many different news site, and we'd need a central module for it too... And many other things. This needs proper thinking

Ok, thank you.
Should we then leave the issue undecided for now?
It will probably take some time to make this plugin releasable anyway :)

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: Fixed. Sorry.
That's what happens when you upload code that you didn't even run...

**Review by Psychokiller1888 — APPROVED** [review]
Will give a try at running it this evening

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (maxbachmann) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: You might want a bit more specific intent names, because names like Play are pretty likely to clash with intent names other modules might want to use aswell

File: PublishedModules/glueckself/mpdclient/mpdclient.install
Comment (maxbachmann) [inline PublishedModules/glueckself/mpdclient/mpdclient.install]: @Psychokiller1888 maintainers are only additional people maintaining the module so the author does not need to appear in that least right?

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (maxbachmann) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: when self._mpd.connect throws an exception self._mpdConnected gets set to true which appears unwanted

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (maxbachmann) [inline PublishedModules/glueckself/mpdclient/mpdclient.py]: I guess the config should be called mpdClient following the other names

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: You're right, thank you. I think it should be enough to prefix them with mpd, right?

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: Fixed, thank you.

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (glueckself) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: Fixed, thank you.

**Comment by glueckself** [comment]
@Psychokiller1888 Thanks, but I expect it'll crash hopelessly. I know now how to install the module, so I can get it to the point where it starts, connects to the server, and enables the intents.
It'll probably need more utterances too before it becomes useful.

Is Alice compatible with snips-satellite? Because if so, then I can run snips-satellite on my laptop and see what happens (I have a VPN home, but a bad Internet connection here).

File: PublishedModules/glueckself/mpdclient/mpdclient.py
Comment (maxbachmann) [reply PublishedModules/glueckself/mpdclient/mpdclient.py]: yeah that should be absolutely fine

**Comment by glueckself** [comment]
I see that there is a ModuleStartingFailed exception. Can I raise it in the constructor or should I use a onStart() method in the module and raise it there? I would raise it if the host is not set or if the password is wrong.

**Comment by Psychokiller1888** [comment]
Yes you can raise it and it won't get loaded at all.

Btw, re configuring module intents. You should use ModuleManager.configureModuleIntents instead of what I said on MqttManager, makes it easier ;)

Alice is compatible with satellites, yes, raw snips sats are ok. The code for complete Alice satellite is not yet ready though

**Comment by glueckself** [comment]
Ok, thank you. And thank you for the support and the feedback! :)

If I understand it correctly, ModuleManager.configureModuleIntents disables or enables all intents. My idea was to enable/disable based on the state, e.g. when playing, then stop, next and prev should be enabled and play disabled; and when stopped, the other way around - play enabled, stop next and prev disabled. Because when I messed around with snips, I had problems with short utterances "music" (->turn on) and "music off"(->turn off), I think that would've helped.

On a second though however, I think that leaving them all enabled might provide better interaction. E.g. when the music is off and you say "stop the music", it might be better if Alice can say that there is no music playing/the server is not connected/... instead of not understanding it because the intent is off.

I'd like to keep the first approach until I get it running :)

I have it installed now btw and I'll push the commits soon.

EDIT: Also, I've renamed it from mpdclient to MpdClient to match the other modules.

**Comment by glueckself** [comment]
Also, I've renamed it from mpdclient to MpdClient to match the other modules.

File: PublishedModules/glueckself/MpdClient/MpdClient.py
Comment (Psychokiller1888) [inline PublishedModules/glueckself/MpdClient/MpdClient.py]: Well done, this is the way to go

**Review by Psychokiller1888 — APPROVED** [review]
Looks ok to me

File: PublishedModules/glueckself/MpdClient/MpdClient.py
Comment (glueckself) [reply PublishedModules/glueckself/MpdClient/MpdClient.py]: This doesn't work, Snips NLU has some errors about the intent id but I'm not sure where they are coming from (the intents are listed in the Snips console). Maybe the id is something else, ModuleManager uses `'intentId': intent.justTopic if hasattr(intent, 'justTopic') else intent`

Comment (glueckself) [reply PublishedModules/glueckself/MpdClient/MpdClient.py]: Workaround, I use always-on-intents now.
While the workaround is better for my module, this may be useful for other modules and should be looked into I think (maybe a low priority issue?)

**Comment by glueckself** [comment]
I guess the module is now ready for a final review.

The module is by far not feature complete, but what's there seems to work:
* connecting and reconnecting
* the commands (play, stop, next, previous), at least in English
* state checks (Alice says that the music is already playing when asked to play while MPD is playing)

The travis error seems unrelated to my module:
```
 find . -name '*.py' | xargs mypy --pretty
PublishedModules/Psychokiller1888/DateDayTimeYear/widgets/__init__.py: error:
Duplicate module named 'widgets' (also at
'./PublishedModules/mjlill/InternationalSpaceStation/widgets/__init__.py')
Found 1 error in 1 file (checked 50 source files)
The command "find . -name '*.py' | xargs mypy --pretty" exited with 123.
```
However, running mypy locally from the ProjectAlice root directory also throws lots of errors (`Found 393 errors in 45 files (checked 1 source file)`), so I guess it would make sense to fix the CI error first to see if mypy reports something about MpdClient.

What remains:
* We've talked about the ModuleStartingFailed exception, but I haven't implemented that yet. Should I do it before the merge?
* Do I have to list myself in the maintainers?
* I have a line `self._playbackStatus = (status['state'] == 'play')`. Is that good code or should I rewrite it as if/else?
* Codacy reports a lots of issues in the README.md. Should I fix them?

I've also noted a few TODOs:
* Volume control
* Playlists (not management, but at least "Hey Alice, please play Nirvana" or "... please play <Playlist> name)
* Use snips injections to add artists/titles into the models so that they can be recognized
* Pause when having a dialogue with Alice
* Pause in general :)

Please let me know if you want some of the easier ones (Volume, Pause) before merging.

And last but not least, thank you guys so much for the good feedback and for the great help!

**Comment by Psychokiller1888** [comment]
There's only one issue left now that I have set codacy correctly :)

**Comment by maxbachmann** [comment]
@glueckself you can run the same tests with mypy that travis uses by executing
```
./.travis/mypy.sh
```
and yes the error is unrelated (the output is clearer now aswell)

File: PublishedModules/glueckself/MpdClient/MpdClient.py
Comment (maxbachmann) [inline PublishedModules/glueckself/MpdClient/MpdClient.py]: if not self._playbackStatus:

**Comment by maxbachmann** [comment]
1) in most other modules we only list ourself as author which automatically implies that we are maintainers aswell
2) the line `self._playbackStatus = (status['state'] == 'play')` is completely fine

**Comment by maxbachmann** [comment]
did you modify mpdhelper.py? or is this a unchanged third party library?

**Comment by glueckself** [comment]
> did you modify mpdhelper.py? or is this a unchanged third party library?

Slightly, because it's actually a part of the source of another mpd client (a GUI called Sonata), so I had to remove some imports (and stuff that depend on those) that were in my way. I've noted in the header what I removed. Is this good enough?

**Comment by maxbachmann** [comment]
@glueckself yes thats absolutely fine, just wondered whether you changed it :)

For the missing features you described I would say that it can be merged without them being implemented. What do you thin @Psychokiller1888 
(for the PR I think it should be squashed since it does not require all the commits and they are all from the same author anyways)

**Pull request merged**
